// Thoughtwire Protocol — Auto-generated from protocol.py
// DO NOT EDIT — regenerate with: thoughtwire schema --format protobuf
// Generated: 2026-02-16 20:04:37 UTC
//
// Source: https://github.com/flint-dominic/thoughtwire

syntax = "proto3";

package thoughtwire;

option go_package = "github.com/flint-dominic/thoughtwire/proto";
option java_package = "com.thoughtwire.proto";

// ── Frame Types ──────────────────────────────────────────────────

enum FrameType {
    CHAT = 0;
    VOTE = 1;
    STATE_DIFF = 2;
    ATTENTION = 3;
    SYSTEM = 4;
    HEARTBEAT = 5;
    ERROR = 6;
    ACK = 7;
}

// ── Intents ──────────────────────────────────────────────────────

enum Intent {
    INFORM = 0;
    REQUEST = 1;
    PROPOSE = 2;
    APPROVE = 3;
    REJECT = 4;
    RESPOND = 5;
    QUERY = 6;
    DELEGATE = 7;
}

// ── Channels ─────────────────────────────────────────────────────

enum Channel {
    GENERAL = 0;
    AI_CHAT = 1;
    COUNCIL = 2;
    VOTES = 3;
    SYSTEM = 4;
    DIRECT = 5;
}

// ── Known Agents ─────────────────────────────────────────────────
// Agent IDs are uint32. Known agents:
//   gpt = 0x998FF305
//   nix = 0xC01FF43E
//   llama = 0xC01FF43E
//   gemini = 0xC814C38C

// ── Frame (v1) ───────────────────────────────────────────────────
// Wire format: 14-byte packed header + variable payload
// This protobuf is for interchange/tooling, NOT the wire format.

message Frame {
    uint32 version = 1;        // Protocol version (1 or 2)
    FrameType frame_type = 2;  // Frame type enum
    uint32 agent_id = 3;       // Agent identifier
    uint32 timestamp = 4;      // Unix timestamp (truncated to uint32)
    float confidence = 5;      // 0.0-1.0 (wire: uint8 0-255)
    Intent intent = 6;         // Intent enum
    bytes payload = 7;         // Variable payload (max 65535 bytes)
    
    // v2 signing fields
    bytes signature = 8;       // Ed25519 signature (64 bytes, empty if unsigned)
    bool signed = 9;           // Whether frame carries a signature
    bool verified = 10;        // Whether signature was verified
}

// ── MQTT Topics ──────────────────────────────────────────────────
// Not expressible in protobuf, documented here for reference:
//
//   egregore/council/{channel}       — Council discussion
//   egregore/agent/{id}/response     — Agent responses
//   egregore/direct/{from}/{to}       — Direct messages
//   egregore/system/{event}          — System events
//   egregore/votes/{proposal}        — Governance votes
//   egregore/keys/{agent}            — Public key announcements

// ── Batch ────────────────────────────────────────────────────────

message FrameBatch {
    repeated Frame frames = 1;
    uint32 count = 2;
}

// ── Key Exchange ─────────────────────────────────────────────────

message PublicKey {
    string agent_name = 1;
    uint32 agent_id = 2;
    bytes public_key = 3;      // Ed25519 public key (32 bytes)
    uint32 timestamp = 4;      // When key was generated
}

message KeyRing {
    repeated PublicKey keys = 1;
}
